Source/iop/Iop_Intrman.cpp

```diff
@@ -165,6 +165,12 @@ uint32 CIntrman::DisableIntrLine(CMIPS& context, uint32 line, uint32 res)
 	mask.f &= ~(1LL << line);
 	context.m_pMemoryMap->SetWord(CIntc::MASK0, mask.h0);
 	context.m_pMemoryMap->SetWord(CIntc::MASK1, mask.h1);
+	if(res != 0)

```

Comment:
So, I've checked some disassembly of `INTRMAN` and it seems that `res` contains the number of the interrupt that was disabled. I think it also can `or` the interrupt number with 0x200 if the interrupt was already off. Probably would benefit from a more in depth check for the `0x200` part, but could you try returning the interrupt number instead of the interrupt state?

Also some info regarding the `res` value here:
https://github.com/ps2dev/ps2sdk/blob/416eb739a261dba5f565584e06db1782c43377f1/iop/kernel/include/intrman.h#L126

Reply:
That header is what I was going from. It says:
A pointer to a variable that will store the current interrupt status
Which is what the code does. Looking at what the BIOS actually does though is of course better :) I should also extend my functional tests to test this on PCSX2 with a real BIOS ... but that will be after the new year (probably)

Reply:
The Sony docs say:

All interrupts and thread dispatches are disabled.
At the same time, the state that was active immediately before interrupts were disabled is stored in the
variable specified by oldstat (use CpuResumeIntr() to return to that state).

Reply:
the real bios does this:

```c++
int DisableIntr(int interrupt, int *res)
{
    int low_irq = interrupt & 0xFF;

    int ictrl;
    CpuSuspendIntr(&ictrl);

    int stat = -103;

    int ret = 0;
    if (low_irq < 0x20) {
        int old_IMASK = IMASK;
        IMASK &= ~(1 << low_irq);

        if (old_IMASK & (1 << low_irq)){
            stat = low_irq;
        } else {
            ret = -103;
        }
    } else if (low_irq < 0x27) {
        if (DMA_ICR & 0xFFFFFF & (1 << (low_irq - 0x10))){
            if (((DMA_ICR & 0xFFFFFF) >> (low_irq - 0x20)) & 0x01){
                stat = low_irq | 0x100;
            }
            if (DMA_ICR2 & (1 << (low_irq - 0x20))){
                stat |= 0x200;
            }
            DMA_ICR2 &= ~(1 << (low_irq - 0x20)) & 0xFFFFFF;
        } else {
            ret = -103;
        }
    } else if (low_irq < 0x2E) {
        if (DMA_ICR2 & 0xFFFFFF & (1 << (low_irq - 0x18))) {
            stat = low_irq;
            if (((DMA_ICR2 & 0xFFFFFF) >> (low_irq - 0x21)) & 0x01){
                stat |= 0x200;
            }
            DMA_ICR2 &= ~(1 << (low_irq - 0x18)) & 0xFFFFFF;
        } else {
            ret = -103;
        }
    }

    if (res){
        *(volatile int *)res = stat;
    }
    CpuResumeIntr(ictrl);
    return ret;
}
```

So it looks more complex than described. for 0-31 it does what it says on the tin. For values greater than that though it does something different and can set the flags at 0x100 and 0x200.
The enable routine has similar code to handle this.
I think DOA passes a value < 0x20 so it's fine. It would be good to find a game which actually passes in a value >= 0x20 so that we can understand what it is doing better.

I can add the correct implementation but it will take a while to write some test code to verify it works like the real bios.